pkg/collector/runner/runner.go

```diff
@@ -131,107 +168,112 @@ func (r *Runner) UpdateNumWorkers(numChecks int64) {
 		desiredNumWorkers = config.MaxNumWorkers
 	}
 
-	// Add workers if we don't have enough for this range
-	added := 0
-	for {
-		if numWorkers >= desiredNumWorkers {
-			break
-		}
-		r.AddWorker()
-		numWorkers++
-		added++
-	}
-	if added > 0 {
-		log.Infof("Added %d workers to runner: now at "+runnerStats.Get("Workers").String()+" workers.", added)
-	}
+	r.ensureMinWorkers(desiredNumWorkers)
 }
 
 // Stop closes the pending channel so all workers will exit their loop and terminate
 // All publishers to the pending channel need to have stopped before Stop is called
 func (r *Runner) Stop() {
-	if atomic.LoadUint32(&r.running) == 0 {
-		log.Debug("Runner already stopped, nothing to do here...")
+	if !atomic.CompareAndSwapUint32(&r.isRunning, 1, 0) {
+		log.Debugf("Runner %d already stopped, nothing to do here...", r.id)
 		return
 	}
 
-	log.Info("Runner is shutting down...")
-
-	close(r.pending)
-	atomic.StoreUint32(&r.running, 0)
+	log.Infof("Runner %d is shutting down...", r.id)
+	close(r.pendingChecksChan)
 
-	// stop checks that are still running
-	r.m.Lock()
-	globalDone := make(chan struct{})
 	wg := sync.WaitGroup{}
 
-	// stop all python subprocesses
-	terminateChecksRunningProcesses()
+	// Stop running checks
+	r.checksTracker.WithRunningChecks(func(runningChecks map[check.ID]check.Check) {
+		// Stop all python subprocesses
+		terminateChecksRunningProcesses()
+
+		for _, c := range runningChecks {
+			// Save the check since the goroutine will reuse the pointer otherwise
+			ch := c
```

Comment:
I might be missing some subtlety here, but instead of doing this, couldn't the closure that runs in a separate goroutine below take `c` a parameter? i.e.:

```
go func(ch check.Check) {
```

Reply:
The approach I used is mostly muscle memory that I always used on these so that it's explicit about the loop variable gotcha. I'll see about changing this.